ark/phenotyping/cluster_test.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+import pandas as pd
+
+import ark.phenotyping.cluster as cluster
+
+
+def test_winner():
+    test_sample = np.array([0.01, 0.02])
+    test_weights = np.array([[[0.001, 0.005],
+                              [0.002, 0.004]],
+                             [[0.003, 0.009],
+                              [0.008, 0.027]]])
+
+    winning_coords = cluster.winner(test_sample, test_weights)
+
+    assert winning_coords == (1, 1)
+
+
+def test_update():
+    test_sample = np.array([0.01, 0.02])
+    test_weights = np.array([[[0.001, 0.005],
+                              [0.002, 0.004]],
+                             [[0.003, 0.009],
+                              [0.008, 0.027]]])
+    test_winning_coords = (1, 1)
+    test_sigma = 1.0
+    test_learning_rate = 0.5
+    test_x_mesh, test_y_mesh = np.meshgrid(np.arange(2), np.arange(2))
+
+    test_t = 1
+    test_num_iters = 1
+
+    # it's impractical to check exact values, so we'll round to make life easier
+    weights = cluster.update(test_sample, test_weights, test_winning_coords,
+                             test_sigma, test_learning_rate, test_x_mesh, test_y_mesh,
+                             test_t, test_num_iters)
+    weights = np.round(weights, decimals=8)
+
+    result = np.array([[[0.00108549, 0.00514249],
+                        [0.00231832, 0.00463663]],
+                       [[0.00327853, 0.00943768],
+                        [0.00833333, 0.02583333]]])
+
+    assert np.all(weights == result)
+
+
+def test_train_som():
```

Comment:
I think we actually can add a correctness check. What about a 2x2 SOM with 3 channels. You can create 4 unique pixel rows, and do 100 iterations or so of training by showing those same 4 pixel rows. Then check that the SOM has converged/is close to those 4 values

Reply:
What's our definition for convergence for this test? Some pixel rows exhibit strong convergence but others are on the order of 3 times different after being trained on a 2x2 SOM with 3 channels (using 4 pixel rows, 100 iterations).

Reply:
I assume right now this runs extremely quickly. If we can keep this entire test under 3 seconds, we can maybe add some more iterations? Or more channels, to help make each unique pixel row more distinct. Perhaps the assertion could be that the difference between each pixel row and its closest neuron is much less (10X?) than the difference between any two neurons?

Whatever you think makes sense.

Reply:
@ngreenwald I can try the other methods first, and if it doesn't work, we can settle for a not-so-specific convergence.

Reply:
So it looks like no matter what we do, we're always going to get a few random weights that is out of range. For the time being, I've added a very flexible check for weight convergence (if no more than the sqrt of the number of weights are out of range).